synapse/storage/state.py

```diff
@@ -185,23 +185,41 @@ def get_state_groups(self, room_id, event_ids):
         })
 
     @defer.inlineCallbacks
-    def _get_state_groups_from_groups(self, groups, types):
-        """Returns dictionary state_group -> (dict of (type, state_key) -> event id)
+    def _get_state_groups_from_groups(self, groups, types, filtered_types=None):
+        """Returns the state groups for a given set of groups, filtering on
+        types of state events.
+
+        Args:
+            groups(list[int]): list of state group IDs to query
+            types(list[str|None, str|None])|None: List of 2-tuples of the form
+                (`type`, `state_key`), where a `state_key` of `None` matches all
+                state_keys for the `type`. If None, all types are returned.
+            filtered_types(list[str]|None): Only apply filtering via `types` to this
```

Comment:
> Only apply filtering via `types`

no, this (and the subsequent implementation) misses the whole point of my suggestion here, which was based around the case of wanting to behave sensibly when `types` ended up empty.

The usecase I was worrying about was:
*  I want to get the list of *all* state events, except for `m.room.member`, which I want to filter by a list of users.
* Due to reasons, that list of users is empty.

So, we have:

```
store._get_state_groups_from_groups(groups, [], filtered_types=['m.room.member'])
```

Your code will return everything here (afaict). 

A correct description of how I envisaged this parameter would be:

> If not None, gives a list of state types which must match entries in `types` to be returned; other types of state are returned unfiltered. If None, *all* state types must match entries in `types`.

(which is still pretty confusing, and I welcome better suggestions. I do think that the current impl is a foot-gun, though.)

Reply:
I seem to have a mental block on this; sorry.  The fact I can't understand what you're proposing makes me wonder how anyone else is going to understand it in future, unless this is just me being particularly thick.

As far as I can tell, my implementation and comment for `filtered_types` matches the behaviour that you're wanting.

```python
store._get_state_groups_from_groups(groups, [], filtered_types=['m.room.member'])
```

gets turned into:

```sql
select stuff where type <> 'm.room.member'
```

...which is what you were asking for?  (types is [], so you don't get any `and type=? and state_key=?` added to the where clause, leaving filtered_types, which is applied as a series of `and type <> ?` clauses).

Is the problem here that the python doc isn't clear enough?  Your suggestion of:

> If not None, gives a list of state types which must match entries in types to be returned; other types of state are returned unfiltered. If None, all state types must match entries in types.

seems to match the behaviour of my suggestion of:

> Only apply filtering via `types` to this list of event types.  Other types of events are returned unfiltered. If None, `types` filtering is applied to all events.

In other words: if `filtered_types` is specified, you should only apply filtering using the `types` field to the types given in the `filtered_types` list.  Others get returned unfiltered.  Is that clearer?

Or have I just got an enormous blindspot on this?

Reply:
> ```python
> store._get_state_groups_from_groups(groups, [], filtered_types=['m.room.member'])
> ```
>
> gets turned into:
>
> ```sql
> select stuff where type <> 'm.room.member'
> ```

Does it? Are you sure? `[]` is falsy, so this block will get skipped: https://github.com/matrix-org/synapse/pull/2970/files#diff-6ee0c965f89a99e084e8500580e3ec17R258. Perhaps you could test this case in the UT.

>  The fact I can't understand what you're proposing makes me wonder how anyone else is going to understand it in future, unless this is just me being particularly thick.

I don't claim that what I'm suggesting is particularly elegant - as I say, I welcome better suggestions. I'm just trying to avoid this landing with the current massive footgun.

I think you just need to change the line I linked above (and its sqlite equivalent) to `if types is not None`.

Reply:
oh, right - i was looking at the sqlite clause which gets it right already with `if types is not None:`; looks like i thinkoed with the `if types:` on the postgres clause. fixed.